File: packages/hquant/src/common/RingDataFrame.ts

// 高性能 DataFrame 实现，支持列式存储和批量操作，基于 TypedRingBuffer
import { TypedRingBuffer } from './TypedRingBuffer';
import { CircularQueue } from './CircularQueue';

export type DataFrameSchema = { [key: string]: 'float' | 'int' | 'string' };
export type DataFrameRow = { [key: string]: number | string };

export class RingDataFrame<T extends DataFrameRow = DataFrameRow> {
  private columns: Map<string, TypedRingBuffer | CircularQueue<string>>;
  private capacity: number;

  constructor(schema: DataFrameSchema, capacity: number) {
    this.columns = new Map();
    this.capacity = capacity;
    for (const key in schema) {
      if (schema[key] === 'string') {
        this.columns.set(key, new CircularQueue<string>(capacity));
      } else {
        this.columns.set(key, new TypedRingBuffer(schema[key] as 'float' | 'int', capacity));
      }
    }
  }

  append(row: T) {
    for (const key of this.columns.keys()) {
      const col = this.columns.get(key)!;
      const value = row[key];
      if (col instanceof CircularQueue) {
        col.push(typeof value === 'string' ? value : String(value ?? ''));
      } else {
        col.push(typeof value === 'number' ? value : NaN);
      }
    }
    // 检查长度是否超限
    if (this.length > this.capacity) throw new Error('DataFrame is full');
  }

  getCol(name: string): TypedRingBuffer | CircularQueue<any> | undefined {
    return this.columns.get(name);
  }

  getRow(index: number): T {
    if (index < 0 || index >= this.length) throw new Error('Index out of bounds');
    const row: any = {};
    for (const [key, col] of this.columns) {
      row[key] = col.get(index);
    }
    return row as T;
  }

  get length(): number {
    // 取第一个列的长度（所有列长度一致）
    const firstCol = this.columns.values().next().value;
    if (!firstCol) return 0;
    return firstCol.size();
  }
  /**
   * 更新指定index的元素
   */
  update(index: number, row: T) {
    if (index < 0 || index >= this.length) throw new Error('Index out of bounds');
    for (const key of this.columns.keys()) {
      const col = this.columns.get(key)!;
      const value = row[key];
      if (col instanceof CircularQueue) {
        col.update(index, typeof value === 'string' ? value : String(value ?? ''));
      } else {
        col.update(index, typeof value === 'number' ? value : NaN);
      }
    }
  }
  toArray(): T[] {
    const arr: T[] = [];
    for (let i = 0; i < this.length; i++) {
      arr.push(this.getRow(i));
    }
    return arr;
  }

  clear() {
    for (const [, col] of this.columns) {
      col.clear();
    }
  }
}

File: packages/hquant-go-v2/common/ring_data_frame.go

package common

import (
	"fmt"
	"sync"
	"time"
)

// DataFrameSchema defines the schema for the DataFrame.
type DataFrameSchema map[string]string // e.g., {"price": "float", "name": "string"}

// RingDataFrame is a column-oriented DataFrame using ring buffers.
type RingDataFrame struct {
	columns  map[string]interface{}
	capacity int
	length   int
	mutex    sync.RWMutex
}

// NewRingDataFrame creates a new RingDataFrame.
func NewRingDataFrame(schema DataFrameSchema, capacity int) (*RingDataFrame, error) {
	if capacity <= 0 {
		return nil, fmt.Errorf("capacity must be positive")
	}

	columns := make(map[string]interface{})
	for key, colType := range schema {
		switch colType {
		case "float", "int":
			columns[key] = NewTypedRingBuffer(capacity)
		case "string":
			columns[key] = NewCircularQueue(capacity)
		case "date":
			columns[key] = NewCircularQueue(capacity)
		default:
			return nil, fmt.Errorf("unsupported column type: %s", colType)
		}
	}

	return &RingDataFrame{
		columns:  columns,
		capacity: capacity,
	}, nil
}

// Push appends a row to the DataFrame.
func (rdf *RingDataFrame) Push(row map[string]interface{}) error {
	rdf.mutex.Lock()
	defer rdf.mutex.Unlock()

	for key, col := range rdf.columns {
		val, ok := row[key]
		if !ok {
			// Handle missing values, e.g., push a zero or empty string
			val = nil // Or a default value based on type
		}

		switch c := col.(type) {
		case *TypedRingBuffer:
			if v, ok := val.(float64); ok {
				c.Push(v)
			} else if v, ok := val.(int); ok {
				c.Push(float64(v))
			} else {
				c.Push(0) // Default for missing/invalid numbers
			}
		case *CircularQueue:
			if v, ok := val.(string); ok {
				c.Push(v)
			} else if v, ok := val.(time.Time); ok {
				c.Push(v)
			} else if val == nil {
				c.Push(nil) // Push nil for missing values
			} else {
				c.Push(fmt.Sprintf("%v", val)) // Convert to string
			}
		}
	}

	if rdf.length < rdf.capacity {
		rdf.length++
	}

	return nil
}

// GetRow returns a row at the specified index.
func (rdf *RingDataFrame) GetRow(index int) (map[string]interface{}, error) {
	rdf.mutex.RLock()
	defer rdf.mutex.RUnlock()

	if index < 0 || index >= rdf.length {
		return nil, fmt.Errorf("index out of bounds")
	}

	row := make(map[string]interface{})
	for key, col := range rdf.columns {
		switch c := col.(type) {
		case *TypedRingBuffer:
			row[key] = c.Get(index)
		case *CircularQueue:
			row[key] = c.Get(index)
		}
	}
	return row, nil
}

// UpdateRow updates a row at the specified index.
func (rdf *RingDataFrame) UpdateRow(index int, row map[string]interface{}) error {
	rdf.mutex.Lock()
	defer rdf.mutex.Unlock()

	if index < 0 || index >= rdf.length {
		return fmt.Errorf("index out of bounds")
	}

	for key, col := range rdf.columns {
		val, ok := row[key]
		if !ok {
			val = nil
		}

		switch c := col.(type) {
		case *TypedRingBuffer:
			if v, ok := val.(float64); ok {
				c.Update(index, v)
			} else if v, ok := val.(int); ok {
				c.Update(index, float64(v))
			} else {
				c.Update(index, 0) // Default for missing/invalid numbers
			}
		case *CircularQueue:
			if v, ok := val.(string); ok {
				c.Update(index, v)
			} else if v, ok := val.(time.Time); ok {
				c.Update(index, v)
			} else if val == nil {
				c.Update(index, nil) // Update with nil
			} else {
				c.Update(index, fmt.Sprintf("%v", val)) // Convert to string
			}
		}
	}
	return nil
}

// Length returns the current number of rows in the DataFrame.
func (rdf *RingDataFrame) Length() int {
	rdf.mutex.RLock()
	defer rdf.mutex.RUnlock()
	return rdf.length
}

// Clear clears the DataFrame.
func (rdf *RingDataFrame) Clear() {
	rdf.mutex.Lock()
	defer rdf.mutex.Unlock()
	for _, col := range rdf.columns {
		switch c := col.(type) {
		case *TypedRingBuffer:
			c.Clear()
		case *CircularQueue:
			c.Clear()
		}
	}
	rdf.length = 0
}
